libnettest2.hpp

```diff
@@ -0,0 +1,1805 @@
+    const std::chrono::time_point<std::chrono::steady_clock> &cbegin = begin;
+    const std::string &ccollector_base_url = collector_base_url;
+    const NettestContext &cctx = ctx;
+    const std::vector<std::string> &cinputs = inputs;
+    const Runner *cthis = this;
+    std::atomic<uint64_t> i{0};
+    std::mutex mutex;
+    const std::string &ctest_start_time = test_start_time;
+    auto pinfo = &info;
+    for (uint8_t j = 0; j < parallelism; ++j) {
+      // Implementation note: make sure this lambda has only access to either
+      // constant stuff or to stuff that it's thread safe.
+      auto main = [
+        &active,               // atomic
+        &cbegin,               // const ref
+        &ccollector_base_url,  // const ref
+        &cctx,                 // const ref
+        &cinputs,              // const ref
+        &ctest_start_time,     // const ref
+        &cthis,                // const pointer
+        &i,                    // atomic
+        &mutex,                // thread safe
+        pinfo                  // ptr to struct w/ only atomic fields
+      ]() noexcept {
+        active += 1;
+        // TODO(bassosimone): more work is required to actually interrupt
+        // "long" tests like NDT that take several seconds to complete. This
+        // is actually broken also in Measurement Kit, where we cannot stop
+        // the reactor easily because of the thread pool. So, it does not
+        // matter much that we're shipping this sub-library with the interrupt
+        // nettest functionality that is not fully functional.
+        while (!cthis->interrupted_) {
+          uint32_t idx = 0;
+          {
+            std::unique_lock<std::mutex> _{mutex};
+            // Implementation note: we currently limit the maximum value of
+            // the index to UINT32_MAX on the grounds that in Java it's painful
+            // to deal with unsigned 64 bit integers.
+            if (i > UINT32_MAX || i >= cinputs.size()) {
+              break;
+            }
+            idx = (uint32_t)i;
+            i += 1;
+          }
+          if (!cthis->run_with_index32(cbegin, ctest_start_time, cinputs, cctx,
+                                       ccollector_base_url, idx, pinfo)) {
+            break;
+          }
+        }
+        active -= 1;
+      };
+      std::thread thread{std::move(main)};
+      thread.detach();
+    }
+    while (active > 0) {
+      constexpr auto msec = 250;
+      std::this_thread::sleep_for(std::chrono::milliseconds(msec));
+    }
+    emit_ev("status.progress", {{"percentage", 0.9},
+                                {"message", "measurement complete"}});
+    // If the report ID is empty, it means we could not open the report for
+    // some reason earlier. In such case, it does not make any sense to attempt
+    // to close a report. It will only create noise in the backend logs.
+    if (!settings_.no_collector && !ctx.report_id.empty()) {
```

Comment:
I am unsure if this is the optimal behaviour, maybe it's better to return an error of report_close also when we failed to open the report.

Reply:
Does it make sense to interpret your suggestion as follows: when the report ID is empty we emit an error signalling that a piece of code, be it submission or closing of a report, has encountered an invalid report ID?

Reply:
The ideal thing, would be that we track if we failed to create a report and if so we also emit an error on report.close, since we failed to close the report due to a missing ID.

Reply:
I do not get why this is preferable than emitting an error when the report cannot be opened, which IIRC is what the code is doing right now. Care to explain?

Reply:
To clarify I believe it would be optimal to do *both*.

I think that given that it failed opening, it is also failing to close it as it doesn't have a reportID. I guess you can say that it's not a failure since it doesn't actually perform any action.

I guess it comes down to a matter of taste.

Reply:
Do we also want to generate failure for submitting the report then? (I am starting to see how generating these failures can be useful as it reduces the amount of state the apps need to keep.)

Reply:
See 5e36ff88bbdb93e0c9deef7ca7b77a6580b5431b